lib/rpc/src/js_tracer.rs

```diff
@@ -0,0 +1,681 @@
+        self.invoke_method("exit", &obj);
+
+        if self.current_depth > 0 {
+            self.current_depth -= 1;
+        }
+    }
+
+    fn on_storage_read(&mut self, _is_transient: bool, address: Address, key: B256, value: B256) {
+        let obj = serde_json::json!({
+            "address": address,
+            "key": key,
+            "value": value,
+        });
+        self.invoke_method("read", &obj);
+    }
+
+    fn on_storage_write(&mut self, _is_transient: bool, address: Address, key: B256, value: B256) {
+        self.storage_overlay
+            .borrow_mut()
+            .insert((address, key), value);
+        let obj = serde_json::json!({
+            "address": address,
+            "key": key,
+            "value": value,
+        });
+
+        self.invoke_method("write", &obj);
+    }
+
+    fn on_bytecode_change(
+        &mut self,
+        address: Address,
+        new_raw_bytecode: Option<&[u8]>,
+        _new_internal_bytecode_hash: B256,
+        new_observable_bytecode_length: u32,
+    ) {
+        if let Some(code) = new_raw_bytecode {
+            let len = new_observable_bytecode_length as usize;
+            let slice = if code.len() >= len {
+                &code[..len]
+            } else {
+                code
+            };
+            let vec = slice.to_vec();
+            self.code_overlay.borrow_mut().insert(address, vec.clone());
+            let obj = serde_json::json!({
+                "address": address,
+                "code": Bytes::from(vec),
+            });
+            self.invoke_method("code", &obj);
+        } else {
+            self.code_overlay.borrow_mut().remove(&address);
+            let obj = serde_json::json!({
+                "address": address,
+                "code": JsonValue::Null,
+            });
+            self.invoke_method("code", &obj);
+        }
+    }
+
+    fn on_event(&mut self, address: Address, topics: Vec<B256>, data: &[u8]) {
+        let obj = serde_json::json!({
+            "address": address,
+            "topics": topics,
+            "data": Bytes::copy_from_slice(data),
+        });
+        self.invoke_method("log", &obj);
+    }
+
+    fn begin_tx(&mut self, _calldata: &[u8]) {
+        self.current_depth = 0;
+        self.pending_create_type = None;
+
+        // Call optional start(config)
+        let config = self.tracer_config.clone();
+        self.invoke_method("start", &config);
+    }
+
+    fn finish_tx(&mut self) {
+        if self.error.is_some() {
+            return;
+        }
+
+        match self.call_result() {
+            Ok(val) => self.results.push(val),
+            Err(err) => self.record_error(err),
+        }
+    }
+
+    fn before_evm_interpreter_execution_step(
+        &mut self,
+        opcode: u8,
+        _frame_state: impl EvmFrameInterface,
+    ) {
+        let obj = serde_json::json!({
+            "op": zk_os_evm_interpreter::opcodes::OPCODE_JUMPMAP[opcode as usize].unwrap_or("Invalid opcode"),
```

Comment:
Also, this is technically supposed to be a JS object with `toString`/`toNumber`/`isPush`